ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -338,46 +340,62 @@ txSubmissionInbound _tracer maxUnacked mpReader mpWriter _version =
+    acknowledgeTxIdsInMempool st txidsSeq _ _ | Seq.null txidsSeq  = st
+    acknowledgeTxIdsInMempool st txidsSeq txidsMap MempoolSnapshot{mempoolHasTx} =
         -- Return the next 'ServerState'
         st {
           availableTxids      = availableTxids',
           bufferedTxs         = bufferedTxs'',
-          unacknowledgedTxIds = unacknowledgedTxIds',
+          unacknowledgedTxIds = unacknowledgedTxIds'',
           numTxsToAcknowledge = numTxsToAcknowledge st
                               + fromIntegral (Seq.length acknowledgedTxIds)
         }
       where
-        -- Divide the available txs in two: those that are already in the
-        -- mempool and those that are not. We'll request some txs from the
+
+        (ignoredTxids, availableTxidsMp) =
+              Map.partitionWithKey
+                (\txid _ -> mempoolHasTx txid)
+                txidsMap
+
+        (_ignoredTxidsU, availableTxidsU) =
+              Map.partitionWithKey
+                (\txid _ -> elem txid (unacknowledgedTxIds st))
+                txidsMap
+
+        -- Divide the new txids in two: those that are already in the
+        -- mempool or in flight and those that are not. We'll request some txs from the
         -- latter.
-        (ignoredTxids, availableTxids') =
-          Map.partitionWithKey
-            (\txid _ -> mempoolHasTx txid)
-            (availableTxids st)
+        availableTxids' = availableTxids st <> Map.intersection availableTxidsMp availableTxidsU
 
         -- The txs that we intentionally don't request, because they are
         -- already in the mempool, need to be acknowledged.
         --
         -- So we extend bufferedTxs with those txs (so of course they have
         -- no corresponding reply).
-        bufferedTxs' = forceElemsToWHNF $
-                       bufferedTxs st
+        bufferedTxs' = bufferedTxs st
                     <> Map.map (const Nothing) ignoredTxids
 
+        unacknowledgedTxIds' = unacknowledgedTxIds st <> txidsSeq
+
         -- Check if having decided not to request more txs we can now
         -- confirm any txids (in strict order in the unacknowledgedTxIds
         -- sequence). This is used in the 'numTxsToAcknowledge' below
         -- which will then be used next time we SendMsgRequestTxIds.
         --
-        (acknowledgedTxIds, unacknowledgedTxIds') =
-          Seq.spanl (`Map.member` bufferedTxs') (unacknowledgedTxIds st)
+        (acknowledgedTxIds, unacknowledgedTxIds'') =
+          Seq.spanl (`Map.member` bufferedTxs') unacknowledgedTxIds'
 
-        -- If so we can remove acknowledged txs from our buffer
-        --
-        bufferedTxs'' = foldl' (flip Map.delete)
+
+        -- If so we can remove acknowledged txs from our buffer provided that they
+        -- are not still in unacknowledgedTxIds''. This happens in case of duplicate
+        -- txids.
+        bufferedTxs'' = forceElemsToWHNF $ foldl' (\m txid -> if elem txid unacknowledgedTxIds''
```

Comment:
Worth creating a set for use when checking membership? As I mentioned elsewhere, I don't have a sense of what the expected sizes will be.

Reply:
The tx windows size is 10 so I don't think a Set is worth it.